### DISCUSSION TITLE
Cannot have more than 1M objects in a table

### DISCUSSION THREAD

**Discussion by rbovard** [QGEP/QGEP#764] [Ideas]
**Describe the bug**
Cannot have more than 1M objects in a table

**To Reproduce**
1. Import large network data
2. See error `ERROR:  nextval: reached maximum value of sequence "seq_reach_point_oid" (999999)`

**Expected behavior**
We can import more than 1M objects in a table

**Desktop (please complete the following information):**
 - QGEP datamodel version: 1.5.6-1

**Comment by sjib** [comment]
This would need an adaption of the following code: https://github.com/QGEP/datamodel/blob/master/02_oid_generation.sql

Currently an OBJ_ID (OID, as STANDARDOID) is generated based on the underlying rules for INTERLIS STANDARDOID with a the prefix of 8 characters  ([see here how to set it in qgep_sys.oid_prefixes](https://qgep.github.io/docs/installation-guide/workstation.html#if-you-are-a-data-owner-add-your-own-oid-to-the-project)) and a postfix of 8 characters. The postfix uses the table shortcut e.g. WN for wastewater_node and the sequence value of the table wn000001, wn000002 etc.

This code could be adapted to generate the sequence part not only with numbers but to use 0..9 and a..z this would increase the range to 36^6 = ca. 2.17 Mia (2'176'782'336) options.

@urskaufmann and @cymed have discussed the issue (and I summarize it here in this discussion) and come to the conclusion that this could be a valid option. They suggest not to use also A..Z (as this would create different sort orders in an export to Excel (not all software handles the order of a..z and A..Z in the same way)

Suggestion is to use the following rule:

wn999999
wn**a**00001
wn**a**00002
etc.

to keep the postfix readable by a human as long as possible.

Their are two options in discussion:

1. To keep the postfix sequence per table or
2. to change it to the whole database so that you would see the order of the objects created over all tables

**Reply by cymed** [reply]
Generally, we can use different enumeration schemes:

Option 1:

wna00009 is followed by wna00010, 
wna99999 is followed by wnaa0000 
wnaaaaaa is followed by wnb00000

+ human readability
+ does not interfere with existing OIDs
- harder to implement

Option 2:

wna00009 is followed by wna0000a, 
wna0000z is followed by wna00010 
wnazzzzz is followed by wnb00000

+ bigger scope for enumeration
+ does not interfere with existing OIDs
- worse readability
- two separate functions (below 1'000'000 and above it)

Option 3:

wn000009 is followed by wn00000a, 
wn00000z is followed by wn000010 

+ easy to implement (convert to base 36)
+ only one function
- interferes with existing OIDs
- even worse readability than 2

**Reply by urskaufmann** [reply]
option 1!

**Reply by cymed** [reply]
I have tried to implement Option 1 for a time now and I am stuck. Help is much appreciated. 

```
CREATE OR REPLACE FUNCTION sys.seq_encode(
	seq bigint)
    RETURNS character varying
    LANGUAGE 'plpgsql'
AS $BODY$
DECLARE
    chars char[];
	charpos bigint[6];	-- 
    ret varchar;		-- retained string
    val bigint;			-- remaining value to be interpreted
BEGIN
    chars := ARRAY['0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z'];
	charpos := ARRAY [0,0,0,0,0,0];
    val := seq; 
    ret := ''; 
    IF val < 10^6 THEN 
        ret := lpad(val::varchar, 6, '0');
	ELSE
		val:=val-10^6;
		charpos[0]=10+val/(36^5)::int; -- set first index, different from all others as 0..9 are already used
		val:=val-(charpos[0]-10)*36^5;
		
		FOR  I IN  REVERSE  5..1 LOOP
		/*
			I struggle with this loop
		*/
		END LOOP;
	END IF;


    RETURN ret;
END;
$BODY$;
```
If we want to maximise the number of inputs, we have to alter Option 1 slightly as i.e. wnaab000 is never reached
Altered Option 1:
wnaaaaaa is not followed by wnb00000, but by wnab0000
